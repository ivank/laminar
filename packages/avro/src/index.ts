/**
 * @packageDocumentation
 * @module @laminar/avro
 */
export { AvroDecimal, bufferToDecimal, decimalToBuffer, to64Bit } from './avro-decimal';
export { AvroTimestampMillis } from './avro-timestamp-millis';
export { AvroEpochDays } from './avro-epoch-days';
