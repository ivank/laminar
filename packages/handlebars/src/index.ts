/**
 * @packageDocumentation
 * @module @laminar/handlebars
 */

import { Middleware } from '@laminar/laminar';
import { readdirSync, readFileSync, existsSync } from 'fs';
import { compile, RuntimeOptions, TemplateDelegate } from 'handlebars';
import { join, relative } from 'path';
import { statSync } from 'fs';

export type Cache = Map<string, { template: TemplateDelegate; mtime: Date }>;

/**
 * Cache types.
 * - preload: load all the templates ones into memory
 * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
 * - none: do not cache templates and load them on every request.
 */
export type CacheType = 'preload' | 'expiry' | 'none';

export interface TemplateConfig {
  /**
   * Helper functions passed to handlebar templates
   */
  helpers?: RuntimeOptions['helpers'];
  /**
   * Main directory where handlebar files should be located, defaults to current working directory.
   * Looks for "partials" and "views" directories inside of it.
   */
  dir?: string;
  /**
   * Directory where all the partials are located, inside the main 'dir' directory, defaults to 'partials'
   */
  partials?: string;
  /**
   * Name of the handlebars file extension. Defaults to 'handlebars', so that only 'my-file.handlebars' kind of files are loaded.
   */
  extension?: string;
  /**
   * Directory name where all the views are located, inside the main 'dir' directory, defaults to 'views'
   */
  views?: string;
  /**
   * Cache types.
   * - preload: load all the templates ones into memory
   * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
   * - none: do not cache templates and load them on every request.
   *
   * Defaults to 'preload'
   */
  cacheType?: CacheType;
  /**
   * Global data to be passed to every template
   */
  globalData?: Record<string, unknown>;
}

export interface CompileTemplatesOptions {
  dir: string;
  extension: string;
}

export type HandlebarsRender = (view: string, data?: Record<string, unknown>) => string;

export interface HandlebarsContext {
  hbs: HandlebarsRender;
}

const hasExtension =
  (extension: string) =>
  (file: string): boolean =>
    file.endsWith(extension);

export const deepReaddirSync = ({ dir, extension }: CompileTemplatesOptions): string[] =>
  readdirSync(dir, { withFileTypes: true })
    .reduce<string[]>((all, item) => {
      return all.concat(
        item.isDirectory() ? deepReaddirSync({ dir: join(dir, item.name), extension }) : join(dir, item.name),
      );
    }, [])
    .filter(hasExtension(extension));

const toName = (dir: string, file: string): string => relative(dir, file.slice(0, file.lastIndexOf('.')));

export const compileTemplates = ({ dir, extension }: CompileTemplatesOptions): Array<[string, TemplateDelegate]> =>
  deepReaddirSync({ dir, extension }).reduce<Array<[string, TemplateDelegate]>>(
    (all, file) => [...all, [toName(dir, file), compile(readFileSync(file, 'utf8'))]],
    [],
  );

export const loadTemplate = (
  name: string,
  cache: Cache,
  cacheType: CacheType,
  { dir, extension }: CompileTemplatesOptions,
): TemplateDelegate => {
  const file = `${join(dir, name)}.${extension}`;

  const stat = cacheType === 'expiry' ? statSync(file) : { mtime: new Date(0) };
  const cached = cache.get(name);

  if (cacheType !== 'none' && cached && stat.mtime < cached.mtime) {
    return cached.template;
  } else {
    const template = compile(readFileSync(file, 'utf8'));
    cache.set(name, { template, mtime: cacheType === 'expiry' ? stat.mtime : new Date() });
    return template;
  }
};

const loadPartials = (options: CompileTemplatesOptions): { [key: string]: TemplateDelegate } =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce((acc, [name, template]) => ({ ...acc, [name]: template }), {})
    : {};

const loadTemplates = (cache: Cache, options: CompileTemplatesOptions): Cache =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce(
        (cache, [name, template]) => cache.set(name, { template, mtime: new Date() }),
        cache,
      )
    : cache;

export const handlebars = ({
  helpers = {},
  dir = process.cwd(),
  extension = 'handlebars',
  partials = 'partials',
  views = 'views',
  cacheType = 'preload',
  globalData = {},
}: TemplateConfig): HandlebarsRender => {
  const cache = cacheType === 'preload' ? loadTemplates(new Map(), { dir, extension }) : new Map();
  const partialTemplates = cacheType === 'preload' ? loadPartials({ dir: join(dir, partials), extension }) : undefined;

  return (name: string, data: Record<string, unknown> = {}) => {
    const template = loadTemplate(name, cache, cacheType, { dir: join(dir, views), extension });
    return template(
      { ...globalData, ...data },
      { helpers, partials: partialTemplates ?? loadPartials({ dir: join(dir, partials), extension }) },
    );
  };
};

export const handlebarsMiddleware = (config: TemplateConfig): Middleware<HandlebarsContext> => {
  const hbs = handlebars(config);
  return (next) => (req) => next({ ...req, hbs });
};
