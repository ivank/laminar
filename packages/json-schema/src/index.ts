/**
 * @packageDocumentation
 * @module @laminar/json-schema
 */

export { ValidationError } from './ValidationError';
export { ResolveError } from './ResolveError';
export {
  InvalidCode,
  Invalid,
  Messages,
  Options,
  Validator,
  Validation,
  error,
  empty,
  onlyErrors,
  errors,
  evaluateItem,
  evaluateProperty,
  skipRest,
  hasErrors,
  hasEvaluatedItem,
  hasEvaluatedProperty,
  combine,
  combineErrors,
  childOptions,
  validateSchema,
} from './validation';
export { Schema, JsonSchema, PrimitiveType } from './schema';
export { toMessages, toMessage, messages } from './messages';
export {
  isCompiled,
  withinContext,
  toSchemaObject,
  compile,
  ResolvedSchema,
  resolve,
  resolveFile,
  RefMap,
  FileContext,
} from './resolve';
export {
  validate,
  validateCompiled,
  ensureValid,
  ensureValidCompiled,
  ValidateOptions,
  ValidateSchemaOptions,
  ValidateCompiledOptions,
  Result,
  ResultError,
  ResultSuccess,
  Drafts,
  FormatErrors,
} from './validate';
export { CoerceCompiledOptions, CoerceSchemaOptions, CoerceOptions, coerce, coerceCompiled } from './coerce';
export { coerceSchema, coercers, CoercerOptions, Coercer } from './coercion';
