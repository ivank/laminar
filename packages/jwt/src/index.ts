/**
 * @packageDocumentation
 * @module @laminar/jwt
 */

export { RequestAuthInfo, RequestSign, User, Session, JWTData, JWTSign, JWTVerify, VerifyJWTData } from './types';
export { jwkPublicKey, JWKHandlerOptions } from './jwk';
export {
  authMiddleware,
  createSession,
  verifyBearer,
  jwtSecurityResolver,
  verifyToken,
  createSessionMiddleware,
  verifyJWT,
} from './jwt';
export {
  verifyKeycloack,
  keycloakAuthMiddleware,
  keycloakJwtSecurityResolver,
  JWTDataKeycloak,
  KeycloakOptions,
} from './keycloak';
