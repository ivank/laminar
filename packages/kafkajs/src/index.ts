/**
 * @packageDocumentation
 * @module @laminarjs/kafkajs
 */
export { KafkaConsumerService } from './kafka-consumer.service';
export { KafkaConsumerOptionalService } from './kafka-consumer-optional.service';
export {
  KafkaProducerService,
  RegisterSchemas,
  producerMiddleware,
  ProducerContext,
  registerSchemas,
  toProducerRecord,
  ToProducerRecordConfig,
  RegisterSchemasConfig,
} from './kafka-producer.service';
export { kafkaLogCreator, produce, Produce } from './helpers';
export {
  DecodedKafkaMessage,
  DecodedEachMessagePayload,
  DecodedBatch,
  DecodedEachBatchPayload,
  EachBatchConsumer,
  EachMessageConsumer,
  EncodedMessage,
  EncodedProducerRecord,
  SchemaRegistryConsumerRunConfig,
  EncodedSchemaProducerRecord,
  EachMessageeMiddleware,
  EachBatchMiddleware,
} from './types';
export { KafkaAdminService, CreateTopics } from './kafka-admin.service';
