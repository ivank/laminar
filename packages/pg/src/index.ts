/**
 * @packageDocumentation
 * @module @laminar/pg
 */
export { PgContext, pgMiddleware, PgMiddlewareConfig } from './pg.middleware';
export { PgService } from './pg.service';
export { PgClient, PgClientConfig } from './pg.client';
