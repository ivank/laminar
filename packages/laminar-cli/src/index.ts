import { ResolvedSchema } from '@ovotech/json-schema';
import { printDocument } from '@ovotech/ts-compose';
import { oas31 } from 'openapi3-ts';
import { convertOapi as apiConvert } from './commands/api/convert';
import { convertOapi as axiosConvert } from './commands/axios/convert';
import { convertSchema } from './convert-schema';

export { convertSchema } from './convert-schema';

export const apiContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(
    apiConvert(
      { root: schema as oas31.OpenAPIObject & { [index: string]: unknown }, refs },
      schema as oas31.OpenAPIObject & { [index: string]: unknown },
    ),
  );

export const axiosContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(
    axiosConvert(
      { root: schema as oas31.OpenAPIObject & { [index: string]: unknown }, refs },
      schema as oas31.OpenAPIObject & { [index: string]: unknown },
    ),
  );

export const schemaContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(convertSchema({ root: schema as oas31.SchemaObject, refs }, schema));

export { toTypeScript as avroConvert } from './commands/avro/convert';
export { toTypeScript as fixturesConvert } from './commands/fixtures/convert';
export { toExternalContext as avroExternalConvert } from './commands/avro/convert';
