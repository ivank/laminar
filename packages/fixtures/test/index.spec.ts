import { Client } from 'pg';
import { generate, setUp, tearDown, toSetupQueries, toTeardownQueries, Fixture } from '../src';
import { id, fixture, rel, template, alternate, cloneFixture } from '../src';

describe('Laminar fixtures', () => {
  it('Should start and stop services', async () => {
    const address: Fixture = fixture('addresses', {
      id,
      city: template('City %s'),
      county: template('County %s'),
      country: 'UK',
      postcode: template('BST SW%s'),
      updated_at: null,
      address_line_1: template('Address %s 1'),
      address_line_2: template('Address %s 2'),
      source_system_id: id,
    });

    const billingAddress = cloneFixture(address);
    const secondaryAddress = cloneFixture(address);

    const contact = fixture('contacts', {
      id,
      address_id: rel(address, 'id'),
      title: alternate('Mr', 'Mrs'),
      first_name: template('Name %s'),
      last_name: template('Name %s'),
      email: template('email%s@example.com'),
      phone: '123123',
      updated_at: null,
    });

    const secondaryContact = cloneFixture(contact, { columns: { address_id: rel(secondaryAddress, 'id') } });

    const customer = fixture('customers', {
      id,
      customer_reference: template('customer-%s'),
      primary_contact_id: rel(contact, 'id'),
      secondary_contact_id: rel(secondaryContact, 'id'),
      billing_address_id: rel(billingAddress, 'id'),
      source_system_id: id,
      generator_id: template('11111-%s'),
      updated_at: null,
    });

    const account = fixture('accounts', {
      id,
      customer_id: rel(customer, 'id'),
      state: 'Active',
      payment_plan: 'BACs',
      updated_at: null,
      start_on: new Date('2020-01-01'),
      end_on: null,
    });

    const installationAddress = cloneFixture(address);

    const installation = fixture('installations', {
      id,
      name: template('Installation %s'),
      type: alternate(
        'Retrofit',
        'New build',
        'Standalone',
        'Extension of an existing FiT-accredited installation',
        'None',
      ),
      technology_type: alternate('PV', 'H', 'W', 'AD', 'CHP'),
      property_type: alternate(
        'Domestic',
        'Commercial',
        'Farm',
        'Industrial',
        'Not for profit',
        'School/Education',
        'Other',
      ),
      export_type: alternate('Deemed', 'Metered Export', 'Off Grid', 'PPA'),
      commissioned_on: new Date('2020-01-01'),
      installed_on: new Date('2020-01-01'),
      mcs_reference: '123',
      tic_reference: 1000,
      dnc_reference: 1000,
      export_mpan: '11111111111111111',
      source_system_id: id,
      legacy_fit_db_id: template('FIT%s'),
      supply_mpan: '11111111111111111',
      address_id: rel(installationAddress, 'id'),
    });

    const generationTariff = fixture('tariffs', {
      id,
      code: template('Tariff %s'),
      type: 'Generation',
      source_system_id: id,
    });
    const exportTariff = cloneFixture(generationTariff, { columns: { type: 'Export' } });

    const contract = fixture('contracts', {
      id,
      scheme_type: 'FIT',
      scheme_account_reference: rel(installation, 'legacy_fit_db_id'),
      terms_and_conditions_agreed: true,
      installation_id: rel(installation, 'id'),
      generation_tariff_id: rel(generationTariff, 'id'),
      export_tariff_id: rel(exportTariff, 'id'),
      source_system_id: id,
      contact_id: rel(contact, 'id'),
      account_id: rel(account, 'id'),
    });

    const generationMeter = fixture('meters', {
      id,
      mpan: '11111111',
      shared: false,
      serial_number: '123123123',
      make: template('Make %s'),
      model: template('Model %s'),
      hh_metered: true,
      gsp: '_L',
      source_system_id: id,
    });
    const exportMeter = cloneFixture(generationMeter);

    const generationInstallationMeter = fixture('installation_meters', {
      id,
      meter_type: 'Generation',
      meter_id: rel(generationMeter, 'id'),
      installation_id: rel(installation, 'id'),
      start_on: new Date('2020-01-01'),
      source_system_id: id,
    });

    const exportInstallationMeter = cloneFixture(generationInstallationMeter, {
      columns: {
        meter_type: 'Export',
        meter_id: rel(exportMeter, 'id'),
      },
    });

    const issue = fixture(
      'issues',
      {
        id,
        description: template('Issue %s'),
        account_id: rel(account, 'id'),
        account_identifier: rel(account, 'account_number'),
        comments: '[]',
        created_at: new Date('2021-07-30'),
        type: 'Account Pending',
        state: 'Open',
        reference_type: 'Account',
        reference_id: rel(account, 'id'),
      },
      { serialIndex: 'issues_id_seq' },
    );

    const fixtures = [contract, generationInstallationMeter, exportInstallationMeter, issue];

    const entities = generate(fixtures);
    expect(entities).toMatchSnapshot('Entites');
    expect(toSetupQueries(500, entities)).toMatchSnapshot('Setup Queries');
    expect(toTeardownQueries(500, entities)).toMatchSnapshot('Teardown Queries');

    const db = new Client({
      connectionString: 'postgres://example-admin:example-pass@localhost:5432/example',
    });

    await db.connect();
    try {
      await setUp({ db, fixtures });
      await expect(db.query(`SELECT nextval('meters_id_seq'::regclass)`)).resolves.toMatchObject({
        rows: [{ nextval: '3' }],
      });
      await expect(db.query(`SELECT nextval('issues_id_seq'::regclass)`)).resolves.toMatchObject({
        rows: [{ nextval: '2' }],
      });
      await tearDown({ db, fixtures });
    } finally {
      await db.end();
    }
  });
});
