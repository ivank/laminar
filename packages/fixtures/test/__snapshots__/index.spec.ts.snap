// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Laminar fixtures Should start and stop services: Entites 1`] = `
[
  {
    "columns": {
      "address_line_1": "Address 1 1",
      "address_line_2": "Address 1 2",
      "city": "City 1",
      "country": "UK",
      "county": "County 1",
      "id": 1,
      "postcode": "BST SW1",
      "source_system_id": 1,
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "addresses",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "address_id": 1,
      "commissioned_on": 2020-01-01T00:00:00.000Z,
      "dnc_reference": 1000,
      "export_mpan": "11111111111111111",
      "export_type": "Deemed",
      "id": 1,
      "installed_on": 2020-01-01T00:00:00.000Z,
      "legacy_fit_db_id": "FIT1",
      "mcs_reference": "123",
      "name": "Installation 1",
      "property_type": "Domestic",
      "source_system_id": 1,
      "supply_mpan": "11111111111111111",
      "technology_type": "PV",
      "tic_reference": 1000,
      "type": "Retrofit",
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "installations",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "code": "Tariff 1",
      "id": 1,
      "source_system_id": 1,
      "type": "Generation",
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "tariffs",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "code": "Tariff 2",
      "id": 2,
      "source_system_id": 2,
      "type": "Export",
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "tariffs",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "address_line_1": "Address 2 1",
      "address_line_2": "Address 2 2",
      "city": "City 2",
      "country": "UK",
      "county": "County 2",
      "id": 2,
      "postcode": "BST SW2",
      "source_system_id": 2,
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "addresses",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "address_id": 2,
      "email": "email1@example.com",
      "first_name": "Name 1",
      "id": 1,
      "last_name": "Name 1",
      "phone": "123123",
      "title": "Mr",
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "contacts",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "address_line_1": "Address 3 1",
      "address_line_2": "Address 3 2",
      "city": "City 3",
      "country": "UK",
      "county": "County 3",
      "id": 3,
      "postcode": "BST SW3",
      "source_system_id": 3,
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "addresses",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "address_id": 3,
      "email": "email2@example.com",
      "first_name": "Name 2",
      "id": 2,
      "last_name": "Name 2",
      "phone": "123123",
      "title": "Mrs",
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "contacts",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "address_line_1": "Address 4 1",
      "address_line_2": "Address 4 2",
      "city": "City 4",
      "country": "UK",
      "county": "County 4",
      "id": 4,
      "postcode": "BST SW4",
      "source_system_id": 4,
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "addresses",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "billing_address_id": 4,
      "customer_reference": "customer-1",
      "generator_id": "11111-1",
      "id": 1,
      "primary_contact_id": 1,
      "secondary_contact_id": 2,
      "source_system_id": 1,
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "customers",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "customer_id": 1,
      "end_on": null,
      "id": 1,
      "payment_plan": "BACs",
      "start_on": 2020-01-01T00:00:00.000Z,
      "state": "Active",
      "updated_at": null,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "accounts",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "account_id": 1,
      "contact_id": 1,
      "export_tariff_id": 2,
      "generation_tariff_id": 1,
      "id": 1,
      "installation_id": 1,
      "scheme_account_reference": "FIT1",
      "scheme_type": "FIT",
      "source_system_id": 1,
      "terms_and_conditions_agreed": true,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "contracts",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "gsp": "_L",
      "hh_metered": true,
      "id": 1,
      "make": "Make 1",
      "model": "Model 1",
      "mpan": "11111111",
      "serial_number": "123123123",
      "shared": false,
      "source_system_id": 1,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "meters",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "id": 1,
      "installation_id": 1,
      "meter_id": 1,
      "meter_type": "Generation",
      "source_system_id": 1,
      "start_on": 2020-01-01T00:00:00.000Z,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "installation_meters",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "gsp": "_L",
      "hh_metered": true,
      "id": 2,
      "make": "Make 2",
      "model": "Model 2",
      "mpan": "11111111",
      "serial_number": "123123123",
      "shared": false,
      "source_system_id": 2,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "meters",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "id": 2,
      "installation_id": 1,
      "meter_id": 2,
      "meter_type": "Export",
      "source_system_id": 2,
      "start_on": 2020-01-01T00:00:00.000Z,
    },
    "serialColumn": "id",
    "serialIndex": undefined,
    "table": "installation_meters",
    "updateMaxSerial": true,
  },
  {
    "columns": {
      "account_id": 1,
      "account_identifier": undefined,
      "comments": "[]",
      "created_at": 2021-07-30T00:00:00.000Z,
      "description": "Issue 1",
      "id": 1,
      "reference_id": 1,
      "reference_type": "Account",
      "state": "Open",
      "type": "Account Pending",
    },
    "serialColumn": "id",
    "serialIndex": "issues_id_seq",
    "table": "issues",
    "updateMaxSerial": true,
  },
]
`;

exports[`Laminar fixtures Should start and stop services: Setup Queries 1`] = `
[
  {
    "text": "INSERT INTO "addresses" ("id", "city", "county", "country", "postcode", "updated_at", "address_line_1", "address_line_2", "source_system_id") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9),($10,$11,$12,$13,$14,$15,$16,$17,$18),($19,$20,$21,$22,$23,$24,$25,$26,$27),($28,$29,$30,$31,$32,$33,$34,$35,$36)",
    "values": [
      1,
      "City 1",
      "County 1",
      "UK",
      "BST SW1",
      null,
      "Address 1 1",
      "Address 1 2",
      1,
      2,
      "City 2",
      "County 2",
      "UK",
      "BST SW2",
      null,
      "Address 2 1",
      "Address 2 2",
      2,
      3,
      "City 3",
      "County 3",
      "UK",
      "BST SW3",
      null,
      "Address 3 1",
      "Address 3 2",
      3,
      4,
      "City 4",
      "County 4",
      "UK",
      "BST SW4",
      null,
      "Address 4 1",
      "Address 4 2",
      4,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('addresses', 'id'), coalesce(max("id"), 0)+1 , false) FROM "addresses";",
    "values": [],
  },
  {
    "text": "INSERT INTO "installations" ("id", "name", "type", "technology_type", "property_type", "export_type", "commissioned_on", "installed_on", "mcs_reference", "tic_reference", "dnc_reference", "export_mpan", "source_system_id", "legacy_fit_db_id", "supply_mpan", "address_id") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16)",
    "values": [
      1,
      "Installation 1",
      "Retrofit",
      "PV",
      "Domestic",
      "Deemed",
      2020-01-01T00:00:00.000Z,
      2020-01-01T00:00:00.000Z,
      "123",
      1000,
      1000,
      "11111111111111111",
      1,
      "FIT1",
      "11111111111111111",
      1,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('installations', 'id'), coalesce(max("id"), 0)+1 , false) FROM "installations";",
    "values": [],
  },
  {
    "text": "INSERT INTO "tariffs" ("id", "code", "type", "source_system_id") VALUES ($1,$2,$3,$4),($5,$6,$7,$8)",
    "values": [
      1,
      "Tariff 1",
      "Generation",
      1,
      2,
      "Tariff 2",
      "Export",
      2,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('tariffs', 'id'), coalesce(max("id"), 0)+1 , false) FROM "tariffs";",
    "values": [],
  },
  {
    "text": "INSERT INTO "contacts" ("id", "address_id", "title", "first_name", "last_name", "email", "phone", "updated_at") VALUES ($1,$2,$3,$4,$5,$6,$7,$8),($9,$10,$11,$12,$13,$14,$15,$16)",
    "values": [
      1,
      2,
      "Mr",
      "Name 1",
      "Name 1",
      "email1@example.com",
      "123123",
      null,
      2,
      3,
      "Mrs",
      "Name 2",
      "Name 2",
      "email2@example.com",
      "123123",
      null,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('contacts', 'id'), coalesce(max("id"), 0)+1 , false) FROM "contacts";",
    "values": [],
  },
  {
    "text": "INSERT INTO "customers" ("id", "customer_reference", "primary_contact_id", "secondary_contact_id", "billing_address_id", "source_system_id", "generator_id", "updated_at") VALUES ($1,$2,$3,$4,$5,$6,$7,$8)",
    "values": [
      1,
      "customer-1",
      1,
      2,
      4,
      1,
      "11111-1",
      null,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('customers', 'id'), coalesce(max("id"), 0)+1 , false) FROM "customers";",
    "values": [],
  },
  {
    "text": "INSERT INTO "accounts" ("id", "customer_id", "state", "payment_plan", "updated_at", "start_on", "end_on") VALUES ($1,$2,$3,$4,$5,$6,$7)",
    "values": [
      1,
      1,
      "Active",
      "BACs",
      null,
      2020-01-01T00:00:00.000Z,
      null,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('accounts', 'id'), coalesce(max("id"), 0)+1 , false) FROM "accounts";",
    "values": [],
  },
  {
    "text": "INSERT INTO "contracts" ("id", "scheme_type", "scheme_account_reference", "terms_and_conditions_agreed", "installation_id", "generation_tariff_id", "export_tariff_id", "source_system_id", "contact_id", "account_id") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)",
    "values": [
      1,
      "FIT",
      "FIT1",
      true,
      1,
      1,
      2,
      1,
      1,
      1,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('contracts', 'id'), coalesce(max("id"), 0)+1 , false) FROM "contracts";",
    "values": [],
  },
  {
    "text": "INSERT INTO "meters" ("id", "mpan", "shared", "serial_number", "make", "model", "hh_metered", "gsp", "source_system_id") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9),($10,$11,$12,$13,$14,$15,$16,$17,$18)",
    "values": [
      1,
      "11111111",
      false,
      "123123123",
      "Make 1",
      "Model 1",
      true,
      "_L",
      1,
      2,
      "11111111",
      false,
      "123123123",
      "Make 2",
      "Model 2",
      true,
      "_L",
      2,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('meters', 'id'), coalesce(max("id"), 0)+1 , false) FROM "meters";",
    "values": [],
  },
  {
    "text": "INSERT INTO "installation_meters" ("id", "meter_type", "meter_id", "installation_id", "start_on", "source_system_id") VALUES ($1,$2,$3,$4,$5,$6),($7,$8,$9,$10,$11,$12)",
    "values": [
      1,
      "Generation",
      1,
      1,
      2020-01-01T00:00:00.000Z,
      1,
      2,
      "Export",
      2,
      1,
      2020-01-01T00:00:00.000Z,
      2,
    ],
  },
  {
    "text": "SELECT setval(pg_get_serial_sequence('installation_meters', 'id'), coalesce(max("id"), 0)+1 , false) FROM "installation_meters";",
    "values": [],
  },
  {
    "text": "INSERT INTO "issues" ("id", "description", "account_id", "account_identifier", "comments", "created_at", "type", "state", "reference_type", "reference_id") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)",
    "values": [
      1,
      "Issue 1",
      1,
      null,
      "[]",
      2021-07-30T00:00:00.000Z,
      "Account Pending",
      "Open",
      "Account",
      1,
    ],
  },
  {
    "text": "SELECT setval('issues_id_seq', coalesce(max("id"), 0)+1 , false) FROM "issues";",
    "values": [],
  },
]
`;

exports[`Laminar fixtures Should start and stop services: Teardown Queries 1`] = `
[
  {
    "text": "DELETE FROM "issues" WHERE id IN ($1)",
    "values": [
      1,
    ],
  },
  {
    "text": "DELETE FROM "installation_meters" WHERE id IN ($1, $2)",
    "values": [
      1,
      2,
    ],
  },
  {
    "text": "DELETE FROM "meters" WHERE id IN ($1, $2)",
    "values": [
      1,
      2,
    ],
  },
  {
    "text": "DELETE FROM "contracts" WHERE id IN ($1)",
    "values": [
      1,
    ],
  },
  {
    "text": "DELETE FROM "accounts" WHERE id IN ($1)",
    "values": [
      1,
    ],
  },
  {
    "text": "DELETE FROM "customers" WHERE id IN ($1)",
    "values": [
      1,
    ],
  },
  {
    "text": "DELETE FROM "contacts" WHERE id IN ($1, $2)",
    "values": [
      1,
      2,
    ],
  },
  {
    "text": "DELETE FROM "tariffs" WHERE id IN ($1, $2)",
    "values": [
      1,
      2,
    ],
  },
  {
    "text": "DELETE FROM "installations" WHERE id IN ($1)",
    "values": [
      1,
    ],
  },
  {
    "text": "DELETE FROM "addresses" WHERE id IN ($1, $2, $3, $4)",
    "values": [
      1,
      2,
      3,
      4,
    ],
  },
]
`;
